### PR TITLE
Fig 2.2 book2 | sub_super_gauss_plot.ipynb

### PR DESCRIPTION
**Pull request by TakshPanchal**
### Figure Number

book2 - Fig-2.2

### Images

| Before PR | After PR |
| :-: | :-: |
|    before     |    after    |

### Issue 

#62 

## Steps

- [X]  Latexified the code
- [X]  Jaxified the code
- [X] Added interactive demo  

## Checklist:

- [X] Performed a self-review of the code
- [X] Tested on Google Colab.

### CODE REVIEW

File: notebooks/book2/02/sub_super_gauss_plot.ipynb [hunk 1]
```diff
@@ -0,0 +1,463 @@
+{
```
Comment (nipunbatra) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <p>Legend missing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (nipunbatra) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <p>For all these plots, use set aspect to equal: https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.set_aspect.html</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (nipunbatra) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def draw_gaussian(n):
</pre><p>all these functions can take seed as an argument</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

**Comment by nipunbatra** [comment]
Other comments

1. this PR has two files. We don't need tfp_hetero.. here
2. top and right spines are not removed in the three of the four figures
3. the a), b), c), d) are missing in three of the four figures

File: notebooks/book2/02/sub_super_gauss_plot.ipynb (hunk 1 above)
Comment (karm-patel) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.        plt.plot(x, normal_dist, color="blue", linestyle="solid", label="Gaussian")
</pre><p>Always use <code>plt.figure()</code> before staring new figure, otherwise while saving figures one figure will go in others. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (karm-patel) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #11.        plt.yticks(jnp.arange(0, 0.55, step=0.05))
</pre><p>Due to this yticks looks very messy</p><p>At this point, it seems you are trying to make non-latexified figure more prettier, but Note that our ultimate goal is to make latexified figure neat and beautiful :) </p><p>It is okay if <code>linewidths</code> in <strong>non-latexified figure</strong> is too small or no. of <code>yticks</code> in <strong> non-latexified figure</strong> are less. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (karm-patel) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        plt.scatter(x1, x2, marker=".", color="green", linewidths=0.1)
</pre><p>By seeing latexified figure, <code>linewidths=0.1</code> looks high, since all samples are overlapped with each other, you can decrease it to make prettier. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (karm-patel) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #20.    draw_all_figs(n, x)
</pre><ol><li>Probably you don't need common plot of all figs here, because in original notebook they have saved individual figures, so you should also do the same, otherwise we will not able to give foot-label (a), (b), (c), etc...</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (karm-patel) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: Also, keep the figure name exactly the same as before while saving.

**Comment by TakshPanchal** [comment]
> * this PR has two files. We don't need tfp_hetero.. here

Sorry, that was a mistake from my side. I removed that extra file.

File: notebooks/book2/02/sub_super_gauss_plot.ipynb (hunk 1 above)
Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: Then, Should I remove this function and generate four figs individually?

Comment (karm-patel) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: Yes

Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: > * top and right spines are not removed in the three of the four figures

In the original figure both the top and right spines have ticks, so should I remove the spines completely same as the first one or add ticks?

**Comment by TakshPanchal** [comment]
> * top and right spines are not removed in the three of the four figures

In the original figure both the top and right spines have ticks, so should I remove the spines completely same as the first one or add ticks?

File: notebooks/book2/02/sub_super_gauss_plot.ipynb (hunk 1 above)
Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: okay, I understand the problem

**Comment by nipunbatra** [comment]
> > * top and right spines are not removed in the three of the four figures
> 
> In the original figure both the top and right spines have ticks, so should I remove the spines completely same as the first one or add ticks?

Yes, remove the ticks.

**Comment by TakshPanchal** [comment]
I updated the code as suggested, kindly check the output again.

| Before PR | After PR |
| :-: | :-: |
|before|after_fig_2 2_v2|

File: notebooks/book2/02/sub_super_gauss_plot.ipynb (hunk 1 above)
Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: If I apply an equal aspect ratio to the first figure, I am getting this output
1D_latexified
.

Comment (nipunbatra) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <p>Equal aspect is needed for the final three figures and not for this one :)</p>

File: notebooks/book2/02/sub_super_gauss_plot.ipynb [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
```
Comment (nipunbatra) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    from scipy.stats import uniform, laplace, norm
</pre><p>In the three separate plots you use <code>jax.random.normal</code> etc. Why do we need <code>scipy.stats</code>?You should just use <code>jax.random.*</code> for all the plots in this notebook.     </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: okay

Comment (karm-patel) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <p>Have you understood the use of <code>width_scale_factor</code> here?</p><p>If not I would recommend you to look at code of <code>latexify</code> : https://github.com/probml/probml-utils/blob/main/probml_utils/plotting.py#:~:text=def%20latexify(,)%3A</p><p> </p><p>In short,</p><p>Our Paper width is 6 inches (in)</p><p>when you give   <code>width_scale_factor=1.70</code> it will generate figures having width 6/1.7 = 3.53 in</p><p>Now we are putting two figures in a row,  so it will take 3.53 + 3.53 = 7.1in which is greater than our paper width 6 in and we don't want to compress our figures to make total width = 6. </p><p> </p><p>So now what <code>width_scale_factor</code> should you choose? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

**Comment by nipunbatra** [comment]
@karm-patel Post my final comments, this looks okay to me. You can perhaps take a final glance. 

The figure titles seem very close to the $\LaTeX$ (a), (b), etc. Perhaps something can be done for the same?

**Comment by karm-patel** [comment]
> @karm-patel Post my final comments, this looks okay to me. You can perhaps take a final glance.
> 
> The figure titles seem very close to the LATEX (a), (b), etc. Perhaps something can be done for the same?

Yes Sir, Perhaps we can fix this issue in latex. for example, we can use `\usepackage{subfig}` and the updated figure looks like the below.
image
But using such packages in book, can break things, so we need to think some other hacks, we'll figure it out.

File: notebooks/book2/02/sub_super_gauss_plot.ipynb [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
```
Comment (karm-patel) [inline notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: <p><code>xlabel</code> (use $x$) &amp; <code>ylabel</code> (use $p(x)$) missing  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pyprobml-review/pyprobml-review/pull/63/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

**Comment by karm-patel** [comment]
@TakshPanchal After resolving pending comments, you can create a PR in the main repo, as of now you can use `\usepackage{subfig}` to put space between two rows of figures.

File: notebooks/book2/02/sub_super_gauss_plot.ipynb (hunk 3 above)
Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: Okay, I will add labels to the first figure

File: notebooks/book2/02/sub_super_gauss_plot.ipynb (hunk 2 above)
Comment (TakshPanchal) [reply notebooks/book2/02/sub_super_gauss_plot.ipynb:1]: Okay, I am figuring it out and will share the image

**Comment by TakshPanchal** [comment]
After using `\usepackage{subfig}` package:

image

**Comment by karm-patel** [comment]
I think you are using `fig_height=2`? I have told you `fig_height=1.5` I guess.

**Comment by TakshPanchal** [comment]
> I think you are using `fig_height=2`? I have told you `fig_height=1.5` I guess.

Yes but by using that first fig squeezes.

**Comment by karm-patel** [comment]
> > I think you are using `fig_height=2`? I have told you `fig_height=1.5` I guess.
> 
> Yes but by using that first fig squeezes.

Then we should use another way to fix it, say decrease `fig_width`
There is a strong reason why we are sticking with the same figure height until and unless it is too much necessary. Currently, Dr. Kevin has a limit of  1300 pages and he is already on the limit. Now say we increase 0.5 inches in each figures (170 figures) and due to this 10-15 pages will be wasted. In these spaces, he could add a whole chapter!

**Comment by TakshPanchal** [comment]
image


> I think you are using `fig_height=2`? I have told you `fig_height=1.5` 

I choose `fig_height=1.5` and `width_scale_factor=2` and generated this figure. Legend of figure one is not looking good I guess. I tried different  values of `width_scale_factor` but always legend overlaps the graph
image

**Comment by karm-patel** [comment]
Okay, looks good. Go ahead with PR in main repo.
